enterprise_gateway/services/kernels/remotemanager.py

```diff
@@ -189,6 +192,57 @@ async def start_kernel(self, *args, **kwargs):
         self.parent.kernel_session_manager.create_session(kernel_id, **kwargs)
         return kernel_id
 
+    async def restart_kernel(self, kernel_id):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the kernel.restarting: {kernel.restarting}")
+        if kernel.restarting:  # assuming duplicate request.
+            await self.wait_for_restart_finish(kernel_id, "restart")
+            self.log.info(
+                f"Done with waiting for restart to complete. Current value of kernel.restarting: {kernel.restarting}. "
+                f"Skipping kernel restart."
+            )
+            return
+        self.log.info("Going ahead to process kernel restart request.")
+        try:
+            kernel.restarting = True  # Moved in out of RemoteKernelManager
+            await super().restart_kernel(kernel_id)
+        finally:
+            self.log.debug("Resetting kernel.restarting flag to False.")
+            kernel.restarting = False
+
+    async def shutdown_kernel(self, kernel_id, now=False, restart=False):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the Kernel Restarting: {kernel.restarting}")
+        if kernel.restarting:
+            await self.wait_for_restart_finish(kernel_id, "shutdown")
+        try:
+            await super().shutdown_kernel(kernel_id, now, restart)
+        except KeyError as ke:  # this is hit for multiple shutdown request.
+            self.log.exception(f"Exception while shutting Kernel: {kernel_id}: {ke}")
+
+    async def wait_for_restart_finish(self, kernel_id, action="shutdown"):
+        kernel = self.get_kernel(kernel_id)
+        start_time = float(time.time())  # epoc time
+        timeout = kernel.kernel_launch_timeout
+        poll_time = 1  # TODO can be read from a new config if need be.
+        self.log.info(
+            f"kernel is restarting when {action} request received. Polling every {poll_time} "
+            f"seconds for next {timeout} seconds for kernel '{kernel_id}'"
+            f" to complete its restart."
+        )
+        while kernel.restarting:
+            now = float(time.time())
+            if (now - start_time) > timeout:
+                self.log.info(
+                    f"Wait_Timeout: Existing out of the restart wait loop to {action} kernel."
+                )
```

Comment:
Let's change this to debug...
```suggestion
                self.log.debug(
                    f"Timeout: Exiting restart wait loop in order to {action} kernel '{kernel_id}'."
                )
```

Reply:
I can change this to debug but I think this is an important log line that needs to be visible by default. Pls give this another thought and let me know .

Reply:
Ok.  Leaving at INFO seems ok.  I'm not sure how useful it is other than to perhaps get an idea that a race condition occurred during restart/restart or restart/shutdown.  That said, if there is noise, this might be something for operators to look into.

Reply:
yes @kevin-bates .. this will be helpful to debug any new issue that might arise due to this new code flow.